driver/panda.c

```diff
@@ -0,0 +1,612 @@
+  }
+
+  memcpy(buf, usb_msg, PANDA_USB_TX_BUFF_SIZE);
+
+  usb_fill_bulk_urb(urb, priv->priv_dev->udev,
+		    usb_sndbulkpipe(priv->priv_dev->udev, 3), buf,
+		    PANDA_USB_TX_BUFF_SIZE, panda_usb_write_bulk_callback,
+		    ctx);
+
+  urb->transfer_flags |= URB_NO_TRANSFER_DMA_MAP;
+  usb_anchor_urb(urb, &priv->tx_submitted);
+
+  err = usb_submit_urb(urb, GFP_ATOMIC);
+  if (unlikely(err))
+    goto failed;
+
+  /* Release our reference to this URB, the USB core will eventually free it entirely. */
+  usb_free_urb(urb);
+
+  return 0;
+
+ failed:
+  usb_unanchor_urb(urb);
+  usb_free_coherent(priv->priv_dev->udev, PANDA_USB_TX_BUFF_SIZE, buf, urb->transfer_dma);
+
+  if (err == -ENODEV)
+    netif_device_detach(priv->netdev);
+  else
+    netdev_warn(priv->netdev, "failed tx_urb %d\n", err);
+
+ nomembuf:
+  usb_free_urb(urb);
+
+  return err;
+}
+
+static void panda_usb_process_can_rx(struct panda_dev_priv *priv_dev,
+				     struct panda_usb_can_msg *msg)
+{
+  struct can_frame *cf;
+  struct sk_buff *skb;
+  int bus_num;
+  struct panda_inf_priv *priv_inf;
+  struct net_device_stats *stats;
+
+  bus_num = (msg->bus_dat_len >> 4) & 0xf;
+  priv_inf = panda_get_inf_from_bus_id(priv_dev, bus_num);
+  if(!priv_inf){
+    printk("Got something on an unused interface %d\n", bus_num);
+    return;
+  }
+  printk("Recv bus %d\n", bus_num);
+
+  stats = &priv_inf->netdev->stats;
+  //u16 sid;
+
+  if (!netif_device_present(priv_inf->netdev))
+    return;
+
+  skb = alloc_can_skb(priv_inf->netdev, &cf);
+  if (!skb)
+    return;
+
+  if(msg->rir & PANDA_CAN_EXTENDED){
+    cf->can_id = (msg->rir >> 3) | CAN_EFF_FLAG;
+  }else{
+    cf->can_id = (msg->rir >> 21);
+  }
+
+  // TODO: Handle Remote Frames
+  //if (msg->dlc & MCBA_DLC_RTR_MASK)
+  //  cf->can_id |= CAN_RTR_FLAG;
+
+  cf->can_dlc = get_can_dlc(msg->bus_dat_len & PANDA_DLC_MASK);
+
+  memcpy(cf->data, msg->data, cf->can_dlc);
+
+  stats->rx_packets++;
+  stats->rx_bytes += cf->can_dlc;
+
+  netif_rx(skb);
+}
+
+static void panda_usb_read_int_callback(struct urb *urb)
+{
+  struct panda_dev_priv *priv_dev = urb->context;
+  int retval;
+  int pos = 0;
+  int inf_num;
+
+  switch (urb->status) {
+  case 0: /* success */
+    break;
+  case -ENOENT:
+  case -ESHUTDOWN:
+    return;
+  default:
+    dev_info(priv_dev->dev, "Rx URB aborted (%d)\n", urb->status);
+    goto resubmit_urb;
+  }
+
+  while (pos < urb->actual_length) {
+    struct panda_usb_can_msg *msg;
+
+    if (pos + sizeof(struct panda_usb_can_msg) > urb->actual_length) {
+      dev_err(priv_dev->dev, "format error\n");
+      break;
+    }
+
+    msg = (struct panda_usb_can_msg *)(urb->transfer_buffer + pos);
+
+    panda_usb_process_can_rx(priv_dev, msg);
+
+    pos += sizeof(struct panda_usb_can_msg);
+  }
+
+ resubmit_urb:
+  usb_fill_int_urb(urb, priv_dev->udev,
+		    usb_rcvintpipe(priv_dev->udev, 1),
+		    urb->transfer_buffer, PANDA_USB_RX_BUFF_SIZE,
+		    panda_usb_read_int_callback, priv_dev, 5);
+
```

Comment:
Does URB_NO_TRANSFER_DMA_MAP have to be set here? If not, why is it set for the first usb_fill_int_urb?